design/src/rfcs/rfc0028_serialization_and_deserialization.md

```diff
@@ -0,0 +1,349 @@
+Additionally, it has come to our attention that some crates just doesn't support it, which would hinder users ability to be productive and tie users' hand.
+
+For the reasons described above, we believe that it is crucial to encode them to string and base64 is favourable over other encoding schemes such as base 16, 32, or Ascii85.
+
+- Reason behind the implementation of non-human readable format
+We considered using the same logic for non-human readable format as well.
+However, readable-ness is not necessary for non-human readable format.
+Additionally, non-human readable format tends to emphasize resource efficiency over human-readable format; Base64 encoded string would take up more space, which is not what the users would want.
+
+Thus, we believe that implementing a tailored serialization logic would be beneficial to the users.
+
+
+### DateTime
+`Serialize` and `Deserialize` is not implemented with derive macro.
+For human-readable format, `DateTime` is serialized in RFC-3339 format;
+It expects the value to be in RFC-3339 format when it is Deserialized.
+
+Non-human readable implements `DateTime` as a tuple of `u32` and `i64`; the latter corresponds to `seconds` field and the first is the `seubsecond_nanos`.
+
+- Reason behind the implementation of a human-readable format
+
+For serialization, `DateTime` format already implements a function to encode itself into RFC-3339 format.
+For deserialization, it is possible to accept other formats, we can add this later if we find it reasonable.
+
+- Reason behind the implementation of a non-human readable format
+
+Serializing them as tuples of two integers results in a smaller data size and requires less computing power than any string-based format.
+Tuple will be smaller in size as it does not require tagging like in maps.
+
+### Document
+`Serialize` and `Deserialize` is implemented with derive macro.
+Additionally, it implements container attribute `#[serde(untagged)]`.
+Serde can distinguish each variant without tagging thanks to the difference in each variant's datatypes.
+
+### Number
+`Serialize` and `Deserialize` is implemented with derive macro.
+Additionally, it implements container attribute `#[serde(untagged)]`.
+
+Serde can distinguish each variant without a tag as each variant's content is different.
+
+## Builder Types and Non-Builder Types
+Builder types and non Builder types implement `Serialize` and `Deserialize` with derive macro.
+Derive macro will be implemented behind a feature-gate; Users must enable `serialize` to use `serde::Serialize`, and `deserialize` to use `serde::Deserialize` respectively. Additionally, user must enable `unstable` until the stablization of this RFC.
+
+```rust
+#[allow(missing_docs)] // documentation missing in model
+#[cfg_attr(
+    all(feature = "unstable", feature = "serialize"),
```

Comment:
I would also probably prefer `cfg(aws-sdk-unstable)` along with clear documentation about how to enable it. (Enabled from `.cargo/config.toml` works quite cleanly. Enabled as a flag causes a lot of headaches because it completely breaks conditional compilation.)

But the main thing here is a clear discussion in the RFC. From my perspective:

Pros of cfg:
- it makes it very obvious that "we might break this and take it away at any time"
- harder to opt into
- can't be opted into by a dependency and cause a surprise

Cons of cfg:
- needs to be clearly documented.
- **Can't** be included transitively (e.g. if a third party crate relies on this feature, it could only be compiled with `cfg` flags

Reply:
Reading about [tokio-unstable](https://docs.rs/tokio/latest/tokio/#unstable-features), I think specifying our unstable "knob" as a `rustc` flag as opposed to a Cargo's feature does have a benefit of having end users always opt in explicitly. Feels safer, i.e. fewer foot guns, in the long run.

Reply:
fixed.